src/Symfony/Component/Serializer/Serializer.php

```diff
@@ -165,17 +165,11 @@ public function normalize(mixed $data, string $format = null, array $context = [
         }
 
         if (is_iterable($data)) {
-            if (is_countable($data) && 0 === \count($data)) {
-                switch (true) {
-                    case ($context[AbstractObjectNormalizer::PRESERVE_EMPTY_OBJECTS] ?? false) && \is_object($data):
-                        if (!$data instanceof \ArrayObject) {
-                            trigger_deprecation('symfony/serializer', '5.4', 'Returning empty object of class "%s" from "%s()" is deprecated. This class should extend "ArrayObject".', get_debug_type($data), __METHOD__);
```

Comment:
Re-reading, I think we should return the original object when it extends ArrayObject

Reply:
I disagree. I produce really bad bahavior I explained in your PR.
See the current diff on the test suite to understand. 

```diff
- "g1":{"list":{"list":[]},"settings":[]},"g2":{"list":["greg"],"settings":[]}}';
+ "g1":{"list":{},"settings":[]},"g2":{"list":["greg"],"settings":[]}}';
```

before: There is an inner `list` when the list is empty, but it does not exist when the list is not empty
after: the shape is always the same, whereas the list is empty or not

Reply:
That a behavior change that we'd better not do imho. The type hint doesn't require this.
Or we'd need another deprecation on 5.4, but I don't see why (especially since there is a test about this).

Reply:
I added this test BTW, to show how it behaves. 
I thought it was clear in the previous conversation... 😬 

I can do what you want, but I persist to say it's a bad idea, since the serializer is not doing what people expect!

Reply:
See #42730 as follow up of this discussion.